File: Go/CG/practice/community/river/river.go
package main

import (
	"fmt"
)

func countNum(num int) int {
	//base case
	if num < 10 {
		return num
	} else {
		return num%10 + countNum(num/10)
	}
}

func river(num int) int {
	return num + countNum(num)
}

/*
func river(num int) int {
	var sum int
	var n = num
	for n > 0 {
		sum += n % 10
		n = n / 10
	}
	num += sum
	return num
}
*/
func main() {
	var r1 int
	fmt.Scan(&r1)

	var r2 int
	fmt.Scan(&r2)

	var x = r1
	var y = r2

	var seenX = make(map[int]bool)
	var seenY = make(map[int]bool)

	seenX[x] = true
	seenY[y] = true

	var res int

	for {
		seenX[river(x)] = true
		seenY[river(y)] = true

		if seenX[river(y)] {
			res = river(y)
			break
		} else if seenY[river(x)] {
			res = river(x)
			break
		}
		x = river(x)
		y = river(y)
	}
	fmt.Println(res)
}

File: C/CG/community/river.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

long long river(long long num) {
	long long sum=0;
	long long n = num;
	while (n > 0) {
		sum += n % 10;
		n = n / 10;
	}
	num += sum;
	return num;
}

int main()
{
    long long r1;
    scanf("%lld", &r1);
    long long r2;
    scanf("%lld", &r2);

    while (r1 != r2){
        if (r1 < r2){
            r1=river(r1);
        }else{
            r2=river(r2);    
        }
    }

    printf("%lld\n",r1);

    return 0;
}
